mbf_costmap_nav/src/mbf_costmap_nav/costmap_navigation_server.cpp

```diff
@@ -885,7 +886,12 @@ bool CostmapNavigationServer::callServiceFindValidPose(mbf_msgs::FindValidPose::
   response.pose.pose.position.x = sol.pose.x;
   response.pose.pose.position.y = sol.pose.y;
   response.pose.pose.position.z = 0;
-  response.pose.pose.orientation = tf::createQuaternionMsgFromYaw(sol.pose.theta);
+
+  // if the difference between solution angle and requested angle is less than ANGLE_INCREMENT,
+  // use the requested one to avoid numerical issues
+  response.pose.pose.orientation = angles::shortest_angular_distance(goal.theta, sol.pose.theta) < ANGLE_INCREMENT ?
+                                       request.pose.pose.orientation :
+                                       tf::createQuaternionMsgFromYaw(sol.pose.theta);
```

Comment:
Suggestion: to avoid similar issues in the future, it might make sense to explicitly check here whether the angle tolerance or dist_tolerance are violated (assert and / or return false)? I.e. `std::hypot(dx, dy) <= dist_tolerance` and `abs(getYaw(response) - getYaw(request)) <= angle_tolerance`.

Another numerical issue that might happen is when the `angle_tolerance` is a multiple of `ANGLE_INCREMENT` (but that should be a separate PR). Position related issues seem unlikely, because the search already uses a hypot check before adding to the queue.

Reply:
I have added the check.

hum, I think we can have numerical issues in different places in this case.
For instance, in the search loop that may return earlier since it may not check all increments correctly (missing the last one because it is a multiple and a floating point error could make it slightly below); or the problem could be when we sum up the increments (floating point error in the sum); or in the conversion again itself. 
I think it will be hard to track all possibilities, so I would advise to use a tolerance slightly above/below increment multiples.

Reply:
So I think adding the check makes sense. But also, we should avoid that this check is even necessary here in the code, because the search loop might return, and the solution gets rejected, so the service returns that there was no solution, even though there exists one that the search would have found if it didn't already terminate.

Also, what would `advising the user to use a tolerance slightly above/below increment multiples` mean? The user would have to write code like `if (std::abs(std::remainer(tolerance, increment)) < 1e-6) {increment -= 1e-6;}`?
Rather than telling the user to do this, I think we should just slightly modify the increment in the code here:

https://github.com/magazino/move_base_flex/blob/c2e7619841ab80557461443b857d1c8514e22e97/mbf_costmap_nav/src/mbf_costmap_nav/free_pose_search.cpp#L154-L157

? 

Something like:
```
const double reduced_tol = std::min(config.angle_tolerance - 1e-6, M_PI);
const int num_steps = 1 + std::max(0, std::ceil(reduced_tol / config.angle_increment));
const double increment = reduced_tol / std::max(1, num_steps - 1);
for (int i = 0; i < num_steps; ++num_steps) {
    const std::vector<double> thetas = i == 0 ? std::vector<double>{ pose_2d.theta } :
                                                   std::vector<double>{ pose_2d.theta + i*increment, pose_2d.theta - i*increment };
```

Reply:
Hum, I don't think we can solve floating point issues that way.

What if the user pass: `config.angle_tolerance = config.angle_increment + 1e-6` ?

Then some conversion, or ROS comm serialization, or floating arithmetic, etc could slightly change that number anyway. Then:  

`reduced_tol ~= config.angle_increment;` (slightly below or slightly above) and then
`num_steps = 1` or `num_steps = 2` ?

Reply:
Maybe I should have clarified: In my suggestion above, the `angle_increment` is interpreted as a maximum step size, and the actual step size is below that, chosen such that the `angle_tolerance` is fully covered, without violating it.

Reply:
Got it. However, if I understood it correctly, we will not fully cover the `angle_tolerance`, but remove `1e-6` from it instead: https://godbolt.org/z/1jhbfY619
Therefore, if the user want to fully cover the tolerance, it will need to send `angle_tolerance + 1e-6` 
And the `angle_increment` is not respected anymore, which may be unexpected for the user experience (as can be seen in the example `tol = inc + 2 * 1e-6;` it uses half of the increment)

Reply:
> Got it. However, if I understood it correctly, we will not fully cover the angle_tolerance, but remove 1e-6 from it instead: https://godbolt.org/z/1jhbfY619

yes. That is done on purpose to not violate the tolerance.

I'd also say it coves the tolerance much much better than the current implementation for some scenarios? E.g.

- `tol: 1.5, inc: 0.6`:
    - now: { 0, 0.6, -0.6, 1.2, -1,2}
    - my suggestion: { 0, 0.4999996667, -0.4999996667, 0.9999993333, -0.9999993333, 1.499999, -1.499999} -> this covers the tolerance better
    
- `tol: 1.2, inc: 0.12` (https://godbolt.org/z/hYvxc63z3)
    - now: {0, ..., 1.08, -1.08}
    - my suggestion: { 0, ..., 1.199999, -1.199999} -> this covers the tolerance better

> Therefore, if the user want to fully cover the tolerance, it will need to send angle_tolerance + 1e-6  (I personally don't like this hardcoded value in the API, it is a bit unexpected)

Since we're dealing with numbers only up to `pi`, I think the `1e-6` value could be reduced to about `std::numeric_limits<double>::epsilon * M_PI`.

Either way, the small reduction of the tolerance has practically basically no effect, unlike the current version has in the two examples above (-> [good enough](https://flylib.com/books/en/1.315.1.17/1/))

> And the angle_increment is not respected anymore, which may be unexpected for the user experience (as can be seen in the example tol = inc + 2 * 1e-6; it uses half of the increment)

Yes so as written above, angle_increment is interpreted as a "maximum step size". So to make this clear to the user, I would of course recommend to rename `angle_increment` to `angle_max_step_size` or sth similar.

Reply:
> I'd also say it coves the tolerance much much better than the current implementation for some scenarios

Yes, indeed. My initial idea was to left to the user to plan a step size and tolerance by themselves, but it is better to do it in the API as you suggested. By changing the name of the variable, we suggest that the algorithm will figure out a better way to handle tolerances and increments. thanks